File: src/main/java/com/leet/medium/SplitLinkedListInParts.java
package com.leet.medium;

import com.leet.common.ListNode;

public class SplitLinkedListInParts {

  // 4 3 3
  // 10 size, number of chunks = 3, chunkSize = 10/3 = 3 + remainder 1

  // 15 4 -> 15/4 = 3
  // 15%4 = 3 (3 chunks from beginning need extra values)
  // 4 4 4 3

  int parse(ListNode node, int pos, int chunks, ListNode[] nodes) {
    if (node == null)
      return 1;
    int size = parse(node.next, pos + 1, chunks, nodes);
    int chunkSize = size / chunks;
    int chunksWithExtraElem = size % chunks;
    int boundary = chunksWithExtraElem * (chunkSize + 1);
    int chunkNumber = pos / (chunkSize + 1);
    if (pos >= boundary) {
      chunkNumber = ((pos - boundary) / chunkSize) + chunksWithExtraElem;
    }
    node.next = null;
    if (nodes[chunkNumber] != null)
      nodes[chunkNumber].next = node;
    else
      nodes[chunkNumber] = node;
    return size;
  }

  public ListNode[] splitListToParts(ListNode head, int k) {
    var nodes = new ListNode[k];
    parse(head, 0, k, nodes);
    return nodes;
  }
}

File: src/main/java/com/leet/medium/SplitLinkedListInParts.go
package main

func splitListToParts(head *ListNode, k int) []*ListNode {
	nodes := make([]*ListNode, k)

	// First, calculate the size of the list
	size, node := 0, head
	for node != nil {
		node = node.Next
		size++
	}

	node = head
	pos, batchSize := 0, size/k
	// Elements in batch should be atleast 1
	if batchSize == 0 {
		batchSize = 1
	}

	for pos < k && node != nil {
		// first size%k elements must have an extra element
		items := batchSize
		if size >= k && pos < size%k {
			items += 1
		}

		// Iterate over the nodes and extract first `items` nodes
		prev, cur := node, node
		for cur != nil && items != 0 {
			prev = cur
			cur = cur.Next
			items--
		}

		// Save the head in the array
		nodes[pos] = node
		if prev != nil {
			prev.Next = nil
		}

		node = cur
		pos++
	}
	return nodes
}
